File: project/src/Rpc/RpcCore.cpp
/******************************************************************************
Copyright (C) 2015 zhuyanxie

　　Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is furnished
to do so, subject to the following conditions:

　　The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
******************************************************************************/


#include "RpcCore.h"

#include <mutex>
#include <iostream>

#include "Reflect/MetaDataTraits.h"
#include "Reflect/ReflectInvoke.h"
#include "Serial/Deserializion.h"
#include "Serial/DeserialType.h"

namespace ls
{

static CRpcCore* s_coreInstance = NULL;
static std::mutex s_lock;

CRpcCore* CRpcCore::instance()
{
    if (NULL == s_coreInstance)
    {
        std::lock_guard<std::mutex> g(s_lock);
        if (NULL == s_coreInstance)
        {
            s_coreInstance = new CRpcCore();
        }
    }

    return s_coreInstance;
}

///\brief           增加一个Rpc session
bool CRpcCore::bindSession(const SessionPtr &session)
{
    return m_session = session;
}

///\brief           调用一个RPC调用
bool CRpcCore::call(RpcCallPtr call, RpcCallPtr &retCall, uint32_t timeout)
{
    if (!m_session)
    {
        return false;
    }

    return this->call(call, retCall, timeout);
}

///\brief       调用
bool CRpcCore::call(const RpcCallPtr &call, RpcCallPtr &retCall, uint32_t timeout)
{
    RpcRequestPtr request(new CRpcRequest(call, m_session));
    insertRequest(request);

    if (!request->call(timeout))
    {
        deleteRequest(request->getRequestId());
        return false;
    }

    retCall = request->getCallReturn();
    if (call->m_method == "attach")
    {
        m_objects.addObject(
                Deserializion<long long>()(retCall->m_values[0].c_str()),
                (void*)Deserializion<long long>()(call->m_values[0].c_str()),
                false);
    }
    else if (call->m_method == "detach")
    {
        m_objects.delObject(Deserializion<long long>()(retCall->m_values[0].c_str()));
    }

    deleteRequest(request->getRequestId());
    return true;
}

///\brief           接收到远程RPC调用
void CRpcCore::onRpcCall(RpcCallPtr rpcCall)
{
    /// 响应
    CRpcResponse resp(rpcCall, m_session);

    /// 类
    CMetaTable *table = nullptr;
    CMetaFactory::instance()->getMetaTable(rpcCall->m_cls.c_str(), table);
    if (nullptr == table)
    {
        return;
    }

    /// 特殊操作对象处理
    if (rpcCall->m_method == "new" || rpcCall->m_method == "delete")
    {
        return createObject(rpcCall, resp);
    }

    /// 方法
    OverLoadMethods *methods = nullptr;
    table->getMethods(rpcCall->m_method.c_str(), methods);
    if (nullptr == methods)
    {
        return;
    }

    /// attach的话先改变对象
    if (rpcCall->m_method == "attach" || rpcCall->m_method == "detach")
    {
        attachObject(rpcCall, resp);
    }

    /// 调用
    rpcCall->m_object = (int64_t) m_objects.getObject(rpcCall->getObject());
    for (auto it = methods->begin(); it != methods->end(); ++it)
    {
        if (onRpcCall(*it, rpcCall, resp))
        {
            break;
        }
    }
}

///\brief           接收到远程RPC的回调
void CRpcCore::onRpcReturn(RpcCallPtr retCall)
{
    RpcRequestPtr request = getRequest(retCall->getCallId());
    if (request)
    {
        request->onReturn(retCall);
    }
}

///\brief       记录一个请求
void CRpcCore::insertRequest(const RpcRequestPtr &request)
{
    std::lock_guard<std::mutex> g(m_lock);
    m_requests[request->getRequestId()] = request;
}

///\brief       删除一个请求
void CRpcCore::deleteRequest(int32_t requestId)
{
    std::lock_guard<std::mutex> g(m_lock);
    m_requests.erase(requestId);
}

///\brief       获取一个请求
RpcRequestPtr CRpcCore::getRequest(int32_t requestId)
{
    std::lock_guard<std::mutex> g(m_lock);
    return m_requests.count(requestId) ? m_requests[requestId] : RpcRequestPtr();
}

///\brief       接收到远程RPC调用，调用本地方法/函数
bool CRpcCore::onRpcCall(const MetaMethodData &metaMethod, RpcCallPtr &call,
        CRpcResponse &resp)
{
    if (metaMethod.params.empty())
    {
        return false;
    }

    /// 重载校验
    auto it     = metaMethod.params.begin();
    auto itv    = call->m_values.begin();
    for (++it; it != metaMethod.params.end() && itv != call->m_values.end();
            ++it, ++itv)
    {
        if (*it != deserialMetaType((*itv).c_str()))
        {
            break;
        }
    }
    if (it != metaMethod.params.end() || itv != call->m_values.end())
    {
        return false;
    }

    /// 函数调用
    RpcCallPtr &returnCall = resp.getReturn();

    MetaDataType returnType = *metaMethod.params.begin();
    switch (returnType)
    {
    case MetaDataTypeVoid:
        ReflectMethodInvoker::invoke<void>(metaMethod, call);
        break;
    case MetaDataTypeInt:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                int>(metaMethod, call));
        break;
    case MetaDataTypeLonglong:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                long long>(metaMethod, call));
        break;
    case MetaDataTypeDouble:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                double>(metaMethod, call));
        break;
    case MetaDataTypeString:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                std::string>(metaMethod, call));
        break;
    case MetaDataTypeClass:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                IReflection*>(metaMethod, call));
        break;
    case MetaDataTypeIntList:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                std::list<int> >(metaMethod, call));
        break;
    case MetaDataTypeLonglongList:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                std::list<long long> >(metaMethod, call));
        break;
    case MetaDataTypeDoubleList:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                std::list<double> >(metaMethod, call));
        break;
    case MetaDataTypeStringList:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                std::list<std::string> >(metaMethod, call));
        break;
    case MetaDataTypeClassList:
        returnCall->pushParams(ReflectMethodInvoker::invoke<
                std::list<IReflection*> >(metaMethod, call));
        break;
    default:
        break;
    }

    return true;
}

///\brief           处理对象
void CRpcCore::createObject(RpcCallPtr &call, CRpcResponse &resp)
{
    int64_t id = call->getObject();
    void *object = nullptr;
    if (call->m_method == "new")
    {
        object = CMetaFactory::instance()->createClassByName(
                call->m_cls.c_str());

        if (nullptr != object)
        {
            m_objects.addObject(id, object);
        }
        else
        {
            /// FIXME error info;
        }

        std::cout << "new : " << id << " local : " << object << std::endl;
    }
    else
    {
        object = m_objects.getObject(id);
        if (nullptr != object)
        {
            delete (IReflection* )object;
            m_objects.delObject(id);
        }
        else
        {
            /// FIXME error info;
        }

        std::cout << "delete : " << id << " local : " << object << std::endl;
    }
}

///\brief           处理回调对象
void CRpcCore::attachObject(RpcCallPtr &call, CRpcResponse &resp)
{
    int64_t id = Deserializion<long long>()(call->m_values[0].c_str());
    void *object = (IReflection*)m_objects.getObject(id);

    call->m_values.clear();
    call->pushParams<long long>((long long)(object));

    resp.getReturn()->pushParams<long long>((long long)(object));

    std::cout << "attach : " << id << " local : " << object << 
        " return : " << resp.getReturn()->m_values[0] << std::endl;
}

CRpcCore::CRpcCore()
    : m_server(nullptr)
{
}

CRpcCore::~CRpcCore()
{
}



}

File: project/src/com/LangStack/Rpc/RpcCore.java
package com.LangStack.Rpc;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import com.LangStack.Serial.Deserial;
import com.LangStack.Serial.Serial;
import com.LangStack.Transport.TransSession;

public class RpcCore
{
    private static RpcCore           sCore        = null;
    private TransSession             mSession     = null;
    private Map<Integer, RpcRequest> mRequestPool = null;
    private Lock                     mLock        = null;
    private RpcObjectPool            mObjects     = null;

    /**
     * @return 单例
     */
    public synchronized static RpcCore instance()
    {
        if (sCore == null)
        {
            sCore = new RpcCore();
        }
        return sCore;
    }

    /**
     * @brief   绑定传输会话
     * @param   session     传输会话
     * @return  true/false是否已经启动
     */
    public synchronized boolean bindSession(TransSession session)
    {
        mSession = session;
        return true;
    }

    /**
     * @brief   调用一个RPC调用
     * @param   rpcCall         rpc调用
     * @param   timeout         超时时间
     * @param   o               本地调用对象（attach/detach用）
     * @return  rpc调用返回值
     */
    RpcCall call(RpcCall call, int timeout, Object o)
    {
        /// FIXME 多个session的情况
        RpcRequest req = new RpcRequest(call, mSession);
        insertRequest(req);
        if (!req.call(timeout))
        {
            deleteRequest(req);
            return null;
        }

        RpcCall retCall = req.getCallReturn();
        if (call.mMethod.equals("attach"))
        {
            mObjects.addObject(
                    (Long)Deserial.deserial(retCall.mValues.get(0)), o);
            
            System.out.println("attach : " + (Long)Deserial.deserial(retCall.mValues.get(0)));
        }
        else if (call.mMethod.equals("detach"))
        {
            mObjects.delObject(
                    (Long)Deserial.deserial(retCall.mValues.get(0)));
            
            System.out.println("detach : " + (Long)Deserial.deserial(retCall.mValues.get(0)));
        }

        deleteRequest(req);
        return retCall;
    }

    /**
     * @brief   删除一个请求
     * @param   req   请求
     */
    private void deleteRequest(RpcRequest req)
    {
        mLock.lock();
        mRequestPool.remove(req.getCallId());
        mLock.unlock();
    }


    /**
     * @brief   接收到远程RPC调用
     * @param   call            rpc调用
     */
    public void onRpcCall(RpcCall call)
    {
        RpcResponse resp = new RpcResponse(call, mSession);
        do {
            /// 对象创建销毁特殊处理
            if (call.mMethod.equals("new") || 
                    call.mMethod.equals("delete"))
            {
                createObject(call, resp);
                break;
            }
            
            /// 回调特殊处理
            if (call.mMethod.equals("attach") ||
                    call.mMethod.equals("detach"))
            {
                /// TODO c++一般不会需要java回调
                break;
            }
            
            /// 获取类
            Class<?> c = null;
            try {
                c = Class.forName(call.mClass);
            } catch (ClassNotFoundException e) {
                e.printStackTrace();
                break;
            }
            
            /// 执行方法
            if (!onRpcCall(c, call, resp))
            {
                /// TODO set error code
            }
            
        } while (false);
        
        resp.sendResponse();
    }
    
    /**
     * @brief   执行RPC方法
     * @param   c           java反射类
     * @param   call        RPC调用对象
     * @param   resp        RPC调用返回
     * @return  true/falseRPC调用的非本方法
     */
    private boolean onRpcCall(Class<?> c, RpcCall call, RpcResponse resp)
    {
        try
        {
            Method method = c.getDeclaredMethod(call.mMethod, 
                    call.getValueClassTypes());
            Object o = mObjects.getObject(call.mObject);
            
            try
            {
                Object res = method.invoke(o, call.getValueObject());
                if (method.getReturnType() != void.class)
                {
                    resp.getReturn().pushParam(Serial.serial(res));
                }
                return true;
            } catch (IllegalAccessException e) {
                e.printStackTrace();
            } catch (IllegalArgumentException e) {
                e.printStackTrace();
            } catch (InvocationTargetException e) {
                e.printStackTrace();
            }
        } catch (NoSuchMethodException e) {
            e.printStackTrace();
        } catch (SecurityException e) {
            e.printStackTrace();
        }
        return false;
    }

    /**
     * @brief   创建或销毁本地对象
     * @param       call    远程调用
     * @param       resp    响应
     */
    private void createObject(RpcCall call, RpcResponse resp)
    {
        long id = call.mObject;
        Object o = null;
        
        if (call.mMethod.equals("new"))
        {
            try {
                o = Class.forName(call.mClass).newInstance();
                mObjects.addObject(id, o);
            } catch (InstantiationException e1) {
                e1.printStackTrace();
            } catch (IllegalAccessException e1) {
                e1.printStackTrace();
            } catch (ClassNotFoundException e1) {
                e1.printStackTrace();
            }
        }
        else
        {
            mObjects.delObject(id);
        }
    }

    /**
     * @brief   接收到远程RPC返回
     * @param   rpcCall         rpc调用
     */
    public void onRpcReturn(RpcCall retCall)
    {
        RpcRequest req = getRequest(retCall.mCallId);
        if (req != null)
        {
            req.onReturn(retCall);
        }
    }
    
    /**
     * 构造
     */
    private RpcCore()
    {
        mRequestPool    = new HashMap<Integer, RpcRequest>();
        mLock           = new ReentrantLock();
        mObjects        = new RpcObjectPool();
    }

    /**
     * @brief 记录一个请求
     * @param req   请求
     */
    private void insertRequest(RpcRequest req)
    {
        mLock.lock();
        if (mRequestPool.containsKey(req.getCallId()))
        {
            mRequestPool.remove(req.getCallId());
        }
        mRequestPool.put(req.getCallId(), req);
        mLock.unlock();
    }
    
    /**
     * @brief 获取一个请求
     * @param callId    请求的唯一ID
     * @return
     */
    private RpcRequest getRequest(int callId)
    {
        RpcRequest req = null;
        mLock.lock();
        if (mRequestPool.containsKey(callId))
        {
            req = mRequestPool.get(callId);
        }
        mLock.unlock();
        return req;
    }
}
